instrumentation/kafka-clients/kafka-clients-0.11/library/src/main/java/io/opentelemetry/instrumentation/kafkaclients/KafkaTracing.java

```diff
@@ -0,0 +1,97 @@
+import io.opentelemetry.instrumentation.kafka.KafkaHeadersGetter;
+import io.opentelemetry.instrumentation.kafka.KafkaHeadersSetter;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.header.Headers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KafkaTracing {
+  private static final Logger logger = LoggerFactory.getLogger(KafkaTracing.class);
+
+  private static final TextMapGetter<Headers> GETTER = new KafkaHeadersGetter();
+
+  private static final TextMapSetter<Headers> SETTER = new KafkaHeadersSetter();
+
+  private final Instrumenter<ProducerRecord<?, ?>, Void> producerInstrumenter;
+  private final Instrumenter<ConsumerRecord<?, ?>, Void> consumerProcessInstrumenter;
+
+  KafkaTracing(
+      Instrumenter<ProducerRecord<?, ?>, Void> producerInstrumenter,
+      Instrumenter<ConsumerRecord<?, ?>, Void> consumerProcessInstrumenter) {
+    this.producerInstrumenter = producerInstrumenter;
+    this.consumerProcessInstrumenter = consumerProcessInstrumenter;
+  }
+
+  /** Returns a new {@link KafkaTracing} configured with the given {@link OpenTelemetry}. */
+  public static KafkaTracing create(OpenTelemetry openTelemetry) {
+    return newBuilder(openTelemetry).build();
+  }
+
+  /** Returns a new {@link KafkaTracingBuilder} configured with the given {@link OpenTelemetry}. */
+  public static KafkaTracingBuilder newBuilder(OpenTelemetry openTelemetry) {
+    return new KafkaTracingBuilder(openTelemetry);
+  }
+
+  private static TextMapPropagator propagator() {
+    return GlobalOpenTelemetry.getPropagators().getTextMapPropagator();
+  }
+
+  /**
+   * Build and inject span into record. Return Runnable handle to end the current span.
+   *
+   * @param record the producer record to inject span info.
+   * @return runnable to close the current span
+   */
+  <K, V> Runnable buildAndInjectSpan(ProducerRecord<K, V> record) {
+    Context currentContext = Context.current();
+
+    if (!producerInstrumenter.shouldStart(currentContext, record)) {
+      return () -> {};
+    }
+
+    Context current = producerInstrumenter.start(currentContext, record);
+    try (Scope ignored = current.makeCurrent()) {
+      try {
+        propagator().inject(current, record.headers(), SETTER);
+      } catch (Throwable t) {
+        // it can happen if headers are read only (when record is sent second time)
+        logger.error("failed to inject span context. sending record second time?", t);
+      }
+    }
+
+    return () -> producerInstrumenter.end(current, record, null, null);
+  }
+
+  <K, V> void buildAndFinishSpan(ConsumerRecords<K, V> records) {
+    Context currentContext = Context.current();
+    for (ConsumerRecord<K, V> record : records) {
+      Context linkedContext = propagator().extract(currentContext, record.headers(), GETTER);
```

Comment:
There's no need for a `KafkaHeadersGetter` class; `KafkaConsumerRecordsGetter` could be used instead.